File: web3walletcore/src/commonMain/kotlin/com/sonsofcrypto/web3walletcore/modules/currencyPicker/CurrencyPickerViewModel.kt
package com.sonsofcrypto.web3walletcore.modules.currencyPicker

import com.sonsofcrypto.web3lib.formatters.Formatters
import com.sonsofcrypto.web3walletcore.extensions.Localized

data class CurrencyPickerViewModel(
    val title: String,
    val allowMultipleSelection: Boolean,
    val showAddCustomCurrency: Boolean,
    val sections: List<Section>,
) {
    sealed class Section(
        val name: String,
    ) {
        data class Networks(
            val items: List<Network>
        ): Section(Localized("currencyPicker.networks.title"))

        data class FavouriteCurrencies(
            val items: List<Currency>
        ): Section(Localized("currencyPicker.currencies.favourite.title"))

        data class Currencies(
            val items: List<Currency>
        ): Section(Localized("currencyPicker.currencies.title"))
    }

    data class Network(
        val iconName: String,
        val name: String,
        val isSelected: Boolean,
    )

    data class Currency(
        val id: String,
        val imageName: String,
        val symbol: String,
        val name: String,
        val position: Position,
        val isSelected: Boolean?,
        val tokens: List<Formatters.Output>? = null,
        val fiat: List<Formatters.Output>? = null,
    )

    enum class Position { SINGLE, FIRST, MIDDLE, LAST }
}

File: iosApp/iosApp/Modules/CurrencyPicker/CurrencyPickerViewModel.swift
// Created by web3d4v on 06/06/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import Foundation
import web3walletcore

struct CurrencyPickerViewModel {
    let title: String
    let allowMultiSelection: Bool
    let showAddCustomCurrency: Bool
    let content: Content
}

extension CurrencyPickerViewModel {
    
    enum Content {
        case loading
        case loaded(sections: [Section])
        case error(error: AppsViewModel.Error)
    }

    struct Section {
        let name: String
        let type: `Type`
        let items: [Item]
        
        enum `Type` {
            case networks
            case tokens
        }
    }
    
    enum Item {
        case network(Network)
        case currency(Currency)
    }
    
    struct Network {
        let networkId: String
        let iconName: String
        let name: String
        let isSelected: Bool
    }
    
    struct Currency {
        let id: String
        let imageName: String
        let symbol: String
        let name: String
        let type: CurrencyType
        let position: Position
        
        enum Position {
            case onlyOne
            case first
            case middle
            case last
        }
    }
    
    struct CurrencyType {
        let isSelected: Bool?
        let balance: Balance?
        
        struct Balance {
            let tokens: [Formatters.Output]
            let usdTotal: [Formatters.Output]
        }
    }
}

extension CurrencyPickerViewModel {

    struct Error {
        let title: String
        let body: String
        let actions: [String]
    }
}

extension CurrencyPickerViewModel {

    var sections: [CurrencyPickerViewModel.Section] {
        switch content {
        case let .loaded(sections):
            return sections
        default:
            return []
        }
    }
}
